lib/classes/PluginManager.js

```diff
@@ -167,8 +145,38 @@ class PluginManager {
     return this.asyncPluginInit();
   }
 
+  requireServicePlugin(serviceDir, pluginPath, localPluginsPath) {
```

Comment:
I don't think we need to change this function (and convert it to the method).

It'll be better, to register eventual local plugins in this context: https://github.com/serverless/serverless/blob/ce66591a688ca734eb34eff929163134d1d21ae8/lib/classes/PluginManager.js#L180-L212
Note that there `isLocalPlugin` variable answers wether it's a path we should consider for packaging exclusion

Reply:
Also what's important:
- We should only group there paths to local plugins (so those which are referenced by relative paths) and not all plugins 
- We should detect whether path targets a file or directory, and in case of a directory ensure all _directory_ is excluded. This is slightly tricky to accomplish. Logic I propose is to to compare against CJS resolved path with something as:

```javascript
if (path.extname(pluginPath) === path.extname(pluginResolvedPath)) {
  // assume file, add `pluginResolvedPath` to `localPluginsPaths`
}
if (pluginResolvedPath.slice(0, -path.extname(pluginResolvedPath).length)).endsWith(path.sep + path.basename(pluginPath)) {
  // assume file, add `pluginResolvedPath` to `localPluginsPaths`
}
if (path.dirname(pluginResolvedPath).endsWith(path.sep + pluginPath)) {
  // assume dir, add `path.dirname(pluginResolvedPath)` to `localPluginsPaths`
}
```

Reply:
> We should detect whether path targets a file or directory, and in case of a directory ensure all directory is excluded

@medikoo this approach already does that. Did you check out the `getLocalPluginsPaths()` method and the tests? 

https://github.com/serverless/serverless/pull/10259/files#diff-888782bdfff1d922f13ff8f30784e31610be569c7fb93a8a3bd2611c1808db11R235-R253

Reply:
@shalvah thanks, indeed it does an attempt on that. That's great.

Still I think we need to take into account only relative paths (those that starts with `./` and not all paths), and also we do not need to do `fs.realpathSync` resolution, I think it's more accurate to work with the outcome of `cjsResolve` (it returns fully resolved real paths)

Reply:
> I don't think we need to change this function (and convert it to the method).


Fair, so you're saying we should move the `this.localPluginsPaths` to the calling method? That can work, but we'll need to run `cjsResolve` again to get the correct entry file path so we can figure out whether it's a file or directory. I'm not sure how expensive `cjsResolve(0` is.

Reply:
> Note that there isLocalPlugin variable 

Ah, I just noticed this. But is the variable definition correct? `name.startsWith('./')`. What if a user specifies a path as '../`? Is that supported.

Reply:
> Ah, I just noticed this. But is the variable definition correct? name.startsWith('./'). What if a user specifies a path as '../`? Is that supported?

This would mean that user wants to reach out beyond service folder, we should not support that way (it's safe to ignore)

Reply:
> also we do not need to do fs.realpathSync resolution, I think it's more accurate to work with the outcome of cjsResolve (it returns fully resolved real paths)

They're complementary, not replacements.

For instance, imagine a folder `myplugin` with a `main` entry in `package.json` pointing to `src/main.js`. The user can pass in `./myplugin`, which will work with CJS (resolved to `/path/myplugin/src/main.js`). By checking the realpath of `/myplugin`, we'll get `/path/myplugin`, which we can then compare with the CJS-resolved path for a match. We could probably figure it out somehow using the specified relative path, but it's simpler to just check the realpath.